### PR TITLE
DPL Analysis: grouping extension for empty sets

### PR DESCRIPTION
**Pull request by aalkin**
* Added support for empty underlying tables in soa::Table
* Improved type handling for table slicing kernel
* Temporary removed slicing test in test_ASoA

Now it is possible to not have exact correspondence between grouping
table entries and groups in a grouped table.

### CODE REVIEW

File: Framework/Core/include/Framework/ASoA.h [hunk 1]
```diff
@@ -1107,7 +1057,7 @@ class TableMetadata
   static const o2::framework::expressions::BindingNode _Getter_##Id { _Label_,     \
                                                                       o2::framework::expressions::selectArrowType<_Type_>() }
 
-#define DECLARE_SOA_INDEX_COLUMN(_Name_, _Getter_) DECLARE_SOA_INDEX_COLUMN_FULL(_Name_, _Getter_, int32_t, _Name_##s, "f" #_Name_ "sID")
```
Comment (ktf) [inline Framework/Core/include/Framework/ASoA.h:1110]: By default arrow can store 32bit values max. Why do you need this change?

File: Framework/Core/include/Framework/ASoAHelpers.h [hunk 2]
```diff
@@ -24,6 +24,27 @@
...
+//    throw std::runtime_error("Unable to slice table");
+//  }
+//  splittedTables.reserve(splittedDatums.size());
+//  for (size_t ti = 0; ti < splittedDatums.size(); ++ti) {
+//    auto table = arrow::util::get<std::shared_ptr<arrow::Table>>(splittedDatums[ti].value);
+//    auto offset = offsets[ti];
+//    splittedTables.emplace_back(table, offset);
+//  }
+//  return splittedTables;
+//}
```
Comment (ktf) [inline Framework/Core/include/Framework/ASoAHelpers.h:47]: Why adding it commented out?

File: Framework/Core/test/test_ASoA.cxx [hunk 3]
```diff
@@ -416,32 +416,32 @@ BOOST_AUTO_TEST_CASE(TestConcatTables)
...
+//  rowWriterA(0, 4, 1);
+//  rowWriterA(0, 5, 1);
+//  rowWriterA(0, 6, 1);
+//  rowWriterA(0, 7, 2);
+//  auto tableA = builderA.finalize();
+//  BOOST_REQUIRE_EQUAL(tableA->num_rows(), 8);
+//  using TestA = o2::soa::Table<o2::soa::Index<>, test::X, test::Y>;
+
+//  TestA t = TestA{tableA};
+//  auto s = slice(t, 3, "y");
```
Comment (ktf) [inline Framework/Core/test/test_ASoA.cxx:436]: Why commenting this out?

File: Framework/Core/include/Framework/ASoA.h (hunk 1 above)
Comment (aalkin) [reply Framework/Core/include/Framework/ASoA.h:1110]: Before I completely rework the type handling, I can only infer type from table.num_rows() which returns int64_t. Actual size of index is only known at runtime, unfortunately. Alternatively I can just set everything to int32_t for now...

File: Framework/Core/include/Framework/ASoAHelpers.h (hunk 2 above)
Comment (aalkin) [reply Framework/Core/include/Framework/ASoAHelpers.h:47]: This should be WIP, I just want it to do run the tests before I rename the PR.

**Pull request merged**